src/chemistry/modal_aero/mam_clouds.F90

```diff
@@ -0,0 +1,530 @@
+  use cloud_utilities, only : cldaero_conc_t, cldaero_allocate, cldaero_deallocate, cldaero_uptakerate
+  use modal_aero_data, only : ntot_amode, modeptr_accum, lptr_so4_cw_amode, lptr_msa_cw_amode
+  use modal_aero_data, only : numptrcw_amode, lptr_nh4_cw_amode
+  use modal_aero_data, only : cnst_name_cw, specmw_so4_amode
+  use chem_mods,       only : adv_mass
+  use physconst,       only : gravit
+  use phys_control,    only : phys_getopts, cam_chempkg_is
+  use chem_mods,       only : gas_pcnst
+  use physics_buffer,  only : physics_buffer_desc
+
+  implicit none
+  private
+
+  public :: sox_cldaero_init
+  public :: sox_cldaero_create_obj
+  public :: sox_cldaero_update
+  public :: sox_cldaero_destroy_obj
+
+  integer :: id_msa, id_h2so4, id_so2, id_h2o2, id_nh3
+
+  real(r8), parameter :: small_value = 1.e-20_r8
+
+contains
+
+!----------------------------------------------------------------------------------
+!----------------------------------------------------------------------------------
+
+  subroutine sox_cldaero_init
+
+    integer :: l, m
+    logical :: history_aerosol      ! Output the MAM aerosol tendencies
+
+    id_msa = get_spc_ndx( 'MSA' )
+    id_h2so4 = get_spc_ndx( 'H2SO4' )
+    id_so2 = get_spc_ndx( 'SO2' )
+    id_h2o2 = get_spc_ndx( 'H2O2' )
+    id_nh3 = get_spc_ndx( 'NH3' )
+
+    if (id_h2so4<1 .or. id_so2<1 .or. id_h2o2<1) then
+      call endrun('sox_cldaero_init:MAM mech does not include necessary species' &
+                  //' -- should not invoke sox_cldaero_mod ')
+    endif
+
+    call phys_getopts( history_aerosol_out        = history_aerosol   )
+    !
+    !   add to history
+    !
+
+  end subroutine sox_cldaero_init
+
+!----------------------------------------------------------------------------------
+!----------------------------------------------------------------------------------
+  function sox_cldaero_create_obj(cldfrc, qcw, lwc, cfact, ncol, loffset) result( conc_obj )
+
+    real(r8), intent(in) :: cldfrc(:,:)
+    real(r8), intent(in) :: qcw(:,:,:)
+    real(r8), intent(in) :: lwc(:,:)
+    real(r8), intent(in) :: cfact(:,:)
+    integer,  intent(in) :: ncol
+    integer,  intent(in) :: loffset
+
+    type(cldaero_conc_t), pointer :: conc_obj
+
+
+    integer :: id_so4_1a, id_so4_2a, id_so4_3a, id_so4_4a, id_so4_5a, id_so4_6a
+    integer :: id_nh4_1a, id_nh4_2a, id_nh4_3a, id_nh4_4a, id_nh4_5a, id_nh4_6a
+    integer :: l,n
+    integer :: i,k
+
+    logical :: mode7
+
+    mode7 = ntot_amode == 7
+
+    conc_obj => cldaero_allocate()
+
+    do k = 1,pver
+       do i = 1,ncol
+          if( cldfrc(i,k) >0._r8) then
+             conc_obj%xlwc(i,k) = lwc(i,k) *cfact(i,k) ! cloud water L(water)/L(air)
+             conc_obj%xlwc(i,k) = conc_obj%xlwc(i,k) / cldfrc(i,k) ! liquid water in the cloudy fraction of cell
+          else
+             conc_obj%xlwc(i,k) = 0._r8
+          endif
+       enddo
+    enddo
+
+    conc_obj%no3c(:,:) = 0._r8
+
+    if (mode7) then
```

Comment:
Index gymnastics is confusing here.

Reply:
This indexing can be removed, similar to what is in carma_clouds.F90 Line 72 to 102.